core/src/engine/impl/wait_list.cpp

```diff
@@ -1,98 +1,90 @@
+  using Handle = WaiterQueue::ItemHandle;
 
-constexpr bool kAdopt = false;
+  Impl() = default;
 
-template <class Container, class Value>
-bool IsInIntrusiveContainer(const Container& container, const Value& val) {
-  const auto val_it = Container::s_iterator_to(val);
-  for (auto it = container.cbegin(); it != container.cend(); ++it) {
-    if (it == val_it) {
-      return true;
+  ~Impl() {
+    Waiter waiter;
+    if (waiters_.TryPop(waiter)) {
+      utils::impl::AbortWithStacktrace(
+          "A TaskContext is sleeping in a WaitList that is being destroyed");
     }
   }
 
-  return false;
-}
-
-using MemberHookConfig =
-    boost::intrusive::member_hook<impl::TaskContext,
-                                  impl::TaskContext::WaitListHook,
-                                  &impl::TaskContext::wait_list_hook>;
-
-}  // namespace
-
-struct WaitList::List
-    : public boost::intrusive::make_list<
-          impl::TaskContext, boost::intrusive::constant_time_size<false>,
-          MemberHookConfig>::type {};
+  bool WakeupOne() noexcept {
+    Waiter waiter;
+    while (waiters_.TryPop(waiter)) {
+      if (waiter.context->Wakeup(TaskContext::WakeupSource::kWaitList,
+                                 waiter.epoch)) {
+        return true;
+      }
+    }
+    return false;
+  }
 
-// not implicitly noexcept on focal
-// NOLINTNEXTLINE(hicpp-use-equals-default,modernize-use-equals-default)
-WaitList::WaitList() noexcept {}
+  Handle Append(TaskContext& context) {
+    Waiter waiter;
+    waiter.context.reset(&context, /*add_ref=*/true);
+    waiter.epoch = context.GetEpoch();
+    return waiters_.Push(std::move(waiter));
+  }
 
-WaitList::~WaitList() {
-  UASSERT_MSG(waiting_contexts_->empty(), "Someone is waiting on the WaitList");
-}
+  void Remove(Handle&& handle) noexcept {
+    Waiter waiter;
+    waiters_.Remove(std::move(handle), waiter);
+  }
 
-bool WaitList::IsEmpty(Lock& lock) const noexcept {
-  UASSERT(lock);
-  return waiting_contexts_->empty();
-}
+ private:
+  WaiterQueue waiters_;
+};
 
-void WaitList::Append(
-    Lock& lock, boost::intrusive_ptr<impl::TaskContext> context) noexcept {
-  UASSERT(lock);
-  UASSERT(context);
-  UASSERT_MSG(!context->wait_list_hook.is_linked(), "context already in list");
+WaitList::WaitList() = default;
 
-  waiting_contexts_->push_back(*context.detach());  // referencing, not copying!
-}
+WaitList::~WaitList() = default;
 
-void WaitList::WakeupOne(Lock& lock) {
-  UASSERT(lock);
-  if (!waiting_contexts_->empty()) {
-    boost::intrusive_ptr<impl::TaskContext> context(&waiting_contexts_->front(),
-                                                    kAdopt);
-    context->wait_list_hook.unlink();
+void WaitList::WakeupOne() noexcept { impl_->WakeupOne(); }
 
-    context->Wakeup(impl::TaskContext::WakeupSource::kWaitList,
-                    impl::TaskContext::NoEpoch{});
+void WaitList::WakeupAll() noexcept {
+  while (impl_->WakeupOne()) {
```

Comment:
Random thought: this could potentially be optimized using enqueue_bulk interface of moodycamel